Contrast Checker: check link color

ramonjd opened the pull request:
Resolves #24595

##  Description

Hi! Let's detect the color of an A tag and, if it exists, pass it to the `<ConstrastChecker />`

![Jan-20-2022 15-29-40](https://user-images.githubusercontent.com/6458278/150273739-f213990a-06ae-41e8-beaf-a1f06c7ffa97.gif)

There are WACG guidelines specifically for links:

> For optimal accessibility and usability, maintain the underline on links.  Otherwise two conditions must be met:

> 3:1 contrast between the body text and the link text.
> A "visual cue" (not just a color change) that appears on mouse hover and keyboard focus. The most common way to meet this is to underline the link on hover and focus.

See: https://webaim.org/articles/contrast/#sc141

To keep things straight forward we're applying the same AA standard to links that we use for text. We can make our tests more stringent should the debate take us there.

## How has this been tested?

Insert a Paragraph block, and some nested blocks that all have color controls, for example, a Paragraph block within a Group block.

Play with text, background and alpha colour controls in the Block Inspector sidebar.

👀 Here is some test content containing various scenarios:

<details><summary>Test blocks</summary>

```html
<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#161515"}}},"color":{"background":"#0f0e0e","text":"#f4f0f0"}}} -->
<p class="has-text-color has-background has-link-color" style="background-color:#0f0e0e;color:#f4f0f0">There's no link in this paragraph. The contrast checker should ignore any incoming link colors until there is one.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#231111"}}},"color":{"text":"#090909","background":"#f8f5f5"}}} -->
<p class="has-text-color has-background has-link-color" style="background-color:#f8f5f5;color:#090909">Paragraph text and <a href="https://en.wikipedia.org/wiki/A_Thing_Called_Love_(song)"><em><strong>link with emphasis</strong></em></a> are readable against Paragraph background</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#f4e9e9"}}},"color":{"text":"#090909","background":"#f8f5f5"}}} -->
<p class="has-text-color has-background has-link-color" style="background-color:#f8f5f5;color:#090909">Paragraph text is readable but a <a href="https://en.wikipedia.org/wiki/A_Thing_Called_Love_(song)"><em><strong>link with emphasis</strong></em></a> should trigger a contrast warning against Paragraph background</p>
<!-- /wp:paragraph -->

<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"#f4b507"}}},"color":{"text":"#f6b707","background":"#060606"}}} -->
<div class="wp-block-group has-text-color has-background has-link-color" style="background-color:#060606;color:#f6b707"><!-- wp:columns {"style":{"color":{"background":"#173d3f"}}} -->
<div class="wp-block-columns has-background" style="background-color:#173d3f"><!-- wp:column {"style":{"spacing":{"padding":{"top":"39px","right":"39px","bottom":"39px","left":"39px"}}}} -->
<div class="wp-block-column" style="padding-top:39px;padding-right:39px;padding-bottom:39px;padding-left:39px"><!-- wp:group {"style":{"spacing":{"padding":{"top":"39px","right":"39px","bottom":"39px","left":"39px"}}}} -->
<div class="wp-block-group" style="padding-top:39px;padding-right:39px;padding-bottom:39px;padding-left:39px"><!-- wp:paragraph {"style":{"color":{"text":"#f7f6f6"}}} -->
<p class="has-text-color" style="color:#f7f6f6">Text and <a href="https://en.wikipedia.org/wiki/Ko-Ko_Joe">link</a> are readable against Column background.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"spacing":{"padding":{"top":"22px","right":"22px","bottom":"22px","left":"22px"}},"color":{"background":"#183f42"}}} -->
<div class="wp-block-group has-background" style="background-color:#183f42;padding-top:22px;padding-right:22px;padding-bottom:22px;padding-left:22px"><!-- wp:paragraph {"style":{"color":{"text":"#080707"}}} -->
<p class="has-text-color" style="color:#080707">Paragraph text and Group background should trigger contrast warning.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"color":{"background":"#f56a06d4"}}} -->
<div class="wp-block-group has-background" style="background-color:#f56a06d4"><!-- wp:paragraph {"style":{"color":{"text":"#b9550d"},"elements":{"link":{"color":{"text":"#82532e"}}}}} -->
<p class="has-text-color has-link-color" style="color:#b9550d">Paragraph text and <a href="https://en.wikipedia.org/wiki/The_Bird_(Jerry_Reed_song)">link</a> on transparent Group background shouldn't trigger any warning.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"color":{"background":"#c83333"}}} -->
<div class="wp-block-group has-background" style="background-color:#c83333"><!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#f2efef"}}},"color":{"text":"#ffffffe3"}}} -->
<p class="has-text-color has-link-color" style="color:#ffffffe3">Paragraph transparent text and readable <a href="https://en.wikipedia.org/wiki/The_Bird_(Jerry_Reed_song)">link</a> on Group background should trigger transparency warning.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:social-links {"iconColor":"pale-cyan-blue","iconColorValue":"#8ed1fc","iconBackgroundColor":"pale-cyan-blue","iconBackgroundColorValue":"#8ed1fc","className":"is-style-default"} -->
<ul class="wp-block-social-links has-icon-color has-icon-background-color is-style-default"><!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->

<!-- wp:social-link {"url":"https://wordpress.org","service":"chain"} /--></ul>
<!-- /wp:social-links -->

<!-- wp:navigation {"ref":340,"textColor":"foreground","backgroundColor":"highlight"} /-->

```

</details>

The contrast warning should show when:

<img width="262" alt="Screen Shot 2022-01-24 at 3 25 20 pm" src="https://user-images.githubusercontent.com/6458278/150721962-f44c8b41-3869-4515-a103-b0b7be0f031f.png">

- the background doesn't have transparency AND
- the text or link color has no transparency AND poor contrast in relation to the background

The contrast warning should take priority over the transparency warning. That is, if the text color for example has poor contrast and the link color has transparency, the contrast warning will show.

The transparency warning should show when:

<img width="269" alt="Screen Shot 2022-01-24 at 3 25 36 pm" src="https://user-images.githubusercontent.com/6458278/150721956-47eaa5af-4642-4a62-8644-ade4b19795c6.png">

- the text or link color have decent contrast but one or both have transparency. 
- same as above but also if the background has transparency

Check the text as well. We can pass a description so contrast warnings will specify 'link' if a link triggers a warning. 

<img width="276" alt="Screen Shot 2022-02-05 at 8 15 58 am" src="https://user-images.githubusercontent.com/6458278/152604396-b7e976bc-fa62-4ddf-82f9-8ae72bf24c52.png">

The tests should cover the scenarios as well:

`npm run test-unit packages/block-editor/src/components/contrast-checker/test/index.js`


## Known issues

This is an existing issue. I'm noting here so it doesn't distract from testing :)

The contrast checker will test and warn about unreadable global styles colors even though the user hasn't interacted with the panel. For example, if you have the following color definitions in your theme.json,

```json
	"styles": {
		"color": {
			"background": "#ffffff",
			"text": "#ffffff"
		},
...
```
You'll see the warning as soon as you open the panel:

<img width="279" alt="Screen Shot 2022-01-24 at 3 03 37 pm" src="https://user-images.githubusercontent.com/6458278/150720237-5ae5553a-1c00-4ec9-9ecd-e0e80cae8955.png">

## Follow up work. 

The Social Icons block tests icon  colors and backgrounds. We could test passing an array of colors as a prop, as we did in 00e5e31f402c7231af995afa25a38c42a61d40a3 to get custom messages.

<img width="310" alt="Screen Shot 2022-02-04 at 11 31 31 pm" src="https://user-images.githubusercontent.com/6458278/152604521-fc2ec353-bb26-4c63-95c7-1ce72271ad9e.png">

I reverted 00e5e31f402c7231af995afa25a38c42a61d40a3 because I thought the changes were too sweeping and needed more testing across the app. We would have to update mobile as well.

## Context
Related issue https://github.com/WordPress/gutenberg/issues/37549
Discussed in https://github.com/WordPress/gutenberg/pull/37731#issuecomment-1017065315
Global styles issue found in https://github.com/WordPress/gutenberg/pull/35774#issuecomment-952046182

## Types of changes
Enhancing existing feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

skorasaurus commented:
also would resolve https://github.com/WordPress/gutenberg/issues/24595

ramonjd committed:
Initial commit. Detecting the color of an A tag, if it exists and passing it to the ConstrastChecker

Added base logic and tests.
Very messy. To be optimzized. Don't hate me.

ramonjd commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -64,61 +71,166 @@ function ContrastChecker( {
 
+	// Flag to warn about transparency only if the text is otherwise readable according to colord
+	// to ensure the readability warnings take precedence.
+	let shouldShowTransparencyWarning = false;
```
Please ignore all these conditions above and below for now. 

They're just there to get the logic right first.

ramonjd replied:
It will be a hook eventually.

ramonjd committed:
A WIP commit. Reducing logic complexity into a hook. Will be squished.

ramonjd committed:
WIP to be squashed. Using hook

ramonjd committed:
Reducing logic with passing tests.

ramonjd committed:
Extracting ContrastCheckerMessage component

ramonjd committed:
Extracting readable options

ramonjd commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -6,119 +6,124 @@ import namesPlugin from 'colord/plugins/names';
-	}
-
-	if ( hasTransparency ) {
+	if ( ! textColorHasTransparency && ! linkColorHasTransparency ) {
```
Much of this could probably be a hook of some description, or at least extracted for now. Once we're happy with the way things work, it can happen in another iteration I think.

glendaviesnz replied:
This is testing well for me. I agree that this might benefit from pulling out into some methods to remove some of the `if` nesting, but I struggled to see a quick/easy way of doing this without making it look just as complicated.

ramonjd replied:
Thanks for testing. Yeah, it's fairly convoluted. What's there wasn't my first attempt either 😆 

The alpha rules really throw a pineapple on the pizza in that they're not part of the official "readable" test. Colord treats rgba like rgb.

ramonjd committed:
Update DOM query to find an A tag even if its nested in, say, a strong or em tag.

ramonjd commented on packages/block-editor/src/components/contrast-checker/contrast-checker-message.js:
```diff
@@ -0,0 +1,58 @@
+	if ( shouldShowTransparencyWarning ) {
+		msg = __( 'Transparent text may be hard for people to read.' );
+	} else {
+		const backgroundColorBrightness = colordBackgroundColor.brightness();
```
In another iteration we could maybe make this a cleaner function and remove the message logic, passing the message only.

aaronrobertshaw commented on packages/block-editor/src/components/contrast-checker/contrast-checker-message.js:
```diff
@@ -0,0 +1,58 @@
+			( colordTextColor &&
+				backgroundColorBrightness < colordTextColor.brightness() ) ||
+			( colordLinkColor &&
+				backgroundColorBrightness < colordLinkColor.brightness() )
+				? __(
+						'This color combination may be hard for people to read. Try using a darker background color and/or a brighter text color.'
+				  )
+				: __(
+						'This color combination may be hard for people to read. Try using a brighter background color and/or a darker text color.'
```
I think the messages here could be a little clearer. I can see the reference to text color being a point of confusion when it's actually the link color triggering the warning.

ramonjd replied:
As part of the[ contrast checker message refactor](https://github.com/WordPress/gutenberg/pull/38100#pullrequestreview-861618806) proposed above, we could try to update the messages so that they discern between link and text.

I agree that more clarity is better always.

I'll see how much of the refactor makes sense for this PR, and try to extract points to follow up on.

aaronrobertshaw reviewed (commented):
Thanks for your work on this @ramonjd 🙇 

The code changes make sense to me and it all tests as advertised. The example block content was a very helpful starting point and a nice touch, so kudos for that 👍 

✅ Unit tests pass
✅ Continues to trigger contrast warnings for background and text colors appropriately
✅ Contrast warnings for links work as advertised
✅ Transparency warnings work as advertised also

As discussed by Glen and yourself previously, I think this is in a good place for a little bit of a refactor, then it should be right to land.

I have one small question though regarding the wording of the contrast warning now it also encompasses link colors but whether it warrants tweaking I'll leave up to you.

ramonjd commented:
Thanks a lot for the review @aaronrobertshaw and @glendaviesnz !!

> As discussed by Glen and yourself previously, I think this is in a good place for a little bit of a refactor, then it should be right to land.

👍 I'll try to untangle things. Maybe it's possible to iterate over a set of rules or something a little more readable.

ramonjd committed:
This commit refactors the contrast checker to accept an array of text colors.
This is so it can handle checking multiple text/foreground colors on a single background, e.g., text, link and so on.
Updating arguments and tests.

ramonjd commented:
> Maybe it's possible to iterate over a set of rules or something a little more readable.

Done. Not sure if it's more readable, but it reduces the logic nightmare and opens the checker up to more than two text/foreground colors now.

> I think the messages here could be a little clearer. I can see the reference to text color being a point of confusion when it's actually the link color triggering the warning.

The refactor includes sending custom messages so we can tweak the copy depending on the which text triggers the warning.

ramonjd commented:
Actually I need to wind things back a bit. 

Maybe keep the original props for now as I haven't looked at the mobile files yet.

Also the message component can probably be reintroduced into the main component given its size.

ramonjd committed:
Reverting change in ContrastChecker props model. These things should be done iteratively as it touches too many parts, e.g. mobile.

ramonjd commented:
✅  Maybe keep the original props for now as I haven't looked at the mobile files yet.
✅  Also the message component can probably be reintroduced into the main component given its size.

aaronrobertshaw commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -1,125 +1,139 @@
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import { Notice } from '@wordpress/components';
+
```
Nitpick: This is very minor and could just be a personal preference of mine but I think it makes more sense for the external dependencies to be first, followed by WordPress, then internal. It would also minimise the code changes and mirror the coding guidelines a little more etc.

ramonjd replied:
You're right, thanks for spotting. Late night copy 🍝  that one.

aaronrobertshaw replied:
Figured as much but didn't want to make too many assumptions 🙂

aaronrobertshaw commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -1,125 +1,139 @@
+			break;
+		}
+
+		// If the text color is readable, but transparent, show the transparent warning.
```
I might have missed something but couldn't we arrive here with an unreadable text color that is transparent?

The previous block of code checked if the text color is not readable, then continues if the text or background has transparency.

ramonjd replied:
Yes, we're testing for that scenario. The comment isn't telling the whole truth. I'll update.

aaronrobertshaw commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -1,125 +1,139 @@
+		}
+
+		// If the text color is readable, but transparent, show the transparent warning.
+		if ( textHasTransparency && true === enableAlphaChecker ) {
```
```suggestion
		if ( textHasTransparency && enableAlphaChecker ) {
```

Don't think we need the yoda condition here.

aaronrobertshaw reviewed (approved):
Thanks for continuing to iterate on this @ramonjd 👍 

The approach in keeping the original props makes sense and isolating any refactor to a follow-up sounds good to me. 

The inclusion of this will have a small impact on the PR [switching the color panel to use the `ToolsPanel`](https://github.com/WordPress/gutenberg/pull/34027) as well. I'm only noting that here in case for some reason this PR doesn't land first.

I've left a few minor nitpick comments but I'll approve this in case you'd prefer to dismiss those suggestions.

ramonjd commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -1,125 +1,139 @@
+		return null;
+	}
+
+	const textColors = [
```
This could, in future, be a model for this component's props as illustrated in 00e5e31f402c7231af995afa25a38c42a61d40a3

Then we'd be able to test `n` number of foreground colors against a single background color.

aaronrobertshaw replied:
Agreed, this could form a basis for a future approach. There could be some edge cases though when it comes to hover or focus states etc.

ramonjd committed:
Update packages/block-editor/src/components/contrast-checker/index.js

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

ramonjd committed:
Reorder dependencies
Update comment so that it reflects what's going on in the code.
